### PR TITLE
[Prototype][abandoned] Port Ideogram 4.0 to JAX / MaxDiffusion

### PR DESCRIPTION
**Pull request by mbohlool**
**Title:** Port Ideogram 4.0 to JAX / MaxDiffusion

**Summary:**
This PR ports the Ideogram 4.0 pipeline to JAX. It introduces the native Flax NNX `Ideogram4Transformer`, integrates the `TorchaxQwen3VLTextEncoder` for complex prompt processing, and implements the required asymmetric CFG denoising loop. It also includes caption key re-ordering in the pipeline to enforce Ideogram's strict JSON-based prompt schema (`high_level_description`, `style_description`, `compositional_deconstruction`).

**Generated Examples:**

**Prompt:**
```
{
  "high_level_description": "A sleek red sports car driving through a futuristic cyberpunk city in the rain, with neon reflections on the wet asphalt.",
  "style_description": {
    "aesthetics": "cyberpunk, futuristic, moody, dynamic",
    "lighting": "neon lights, cinematic lighting, reflections",
    "photo": "photorealistic, motion blur on background",
    "medium": "digital photography",
    "color_palette": ["#E74C3C", "#3498DB", "#9B59B6", "#1ABC9C", "#2C3E50"]
  },
  "compositional_deconstruction": {
    "background": "A bustling cyberpunk cityscape at night, towering skyscrapers, heavy rain.",
    "elements": [
      {
        "type": "obj",
        "bbox": [400, 100, 900, 900],
        "desc": "A highly detailed, sleek red futuristic sports car speeding forward."
      },
      {
        "type": "obj",
        "bbox": [800, 0, 1000, 1000],
        "desc": "Wet asphalt road reflecting vibrant pink and blue neon lights."
      }
    ]
  }
}
```

pr_magic_car_ideogram_output_42_0

**Prompt:** 
```
{
  "high_level_description": "A ginger cat wearing a tiny wizard hat reading a glowing spellbook.",
  "style_description": {
    "aesthetics": "magical, fantasy, whimsical",
    "lighting": "warm glowing light, magical illumination from the book",
    "photo": "highly detailed, macro photography, shallow depth of field",
    "medium": "digital photography, fantasy illustration",
    "color_palette": ["#E67E22", "#8E44AD", "#F1C40F", "#2C3E50", "#D35400"]
  },
  "compositional_deconstruction": {
    "background": "A dimly lit, dusty wizard's study filled with old scrolls and potions.",
    "elements": [
      {
        "type": "obj",
        "bbox": [100, 200, 800, 800],
        "desc": "A fluffy ginger cat wearing a detailed purple wizard hat with gold stars."
      },
      {
        "type": "obj",
        "bbox": [700, 200, 1000, 800],
        "desc": "An ancient, open spellbook glowing with magical yellow light."
      }
    ]
  }
}
```
pr_magic_cat_ideogram_output_42_0

**Prompt:**

```
{
  "high_level_description": "A stunning cinematic shot of a futuristic city at night with flying cars and bright neon lights.",
  "style_description": {
    "aesthetics": "cinematic, futuristic, sci-fi",
    "lighting": "neon lighting, night time, glowing cityscapes",
    "photo": "photorealistic, hyper-detailed, wide angle",
    "medium": "digital photography",
    "color_palette": ["#00FFFF", "#FF00FF", "#000080", "#FFD700", "#191970"]
  },
  "compositional_deconstruction": {
    "background": "A dark night sky overlooking a sprawling futuristic megalopolis.",
    "elements": [
      {
        "type": "obj",
        "bbox": [200, 0, 1000, 1000],
        "desc": "Dense cityscape with towering skyscrapers illuminated by intricate neon signs."
      },
      {
        "type": "obj",
        "bbox": [100, 200, 400, 800],
        "desc": "Several futuristic flying cars with glowing thrusters streaking across the sky."
      }
    ]
  }
}
```
pr_magic_city_ideogram_output_42_0

### CODE REVIEW

File: src/maxdiffusion/generate_ideogram.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  max_logging.log("============================================================")
+
+  original_enable_profiler = config.get_keys().get("enable_profiler", False)
+  original_enable_mld = config.get_keys().get("enable_ml_diagnostics", False)
+  original_num_steps = config.get_keys().get("num_inference_steps", 40)
+
+  # 1. Warmup Compilation
+  config.get_keys()["enable_profiler"] = False
+  config.get_keys()["enable_ml_diagnostics"] = False
+  config.get_keys()["num_inference_steps"] = 2  # lower for warmup
```
Comment (eltsai) [inline src/maxdiffusion/generate_ideogram.py]: since we have `z, _, _ = jax.lax.fori_loop(0, num_steps, denoise_step, init_val)` in `src/maxdiffusion/pipelines/ideogram/ideogram_pipeline.py`,  setting the warmup step as 2 means that the later 50 step inference time includes compile


I think this is the reason why we see a 50 sec generation time. We could either:
1. go with what WAN does: only compile one step and do a plain for loop
2. warmup with 50 steps

I am not sure which would be faster, it's worth testing

File: src/maxdiffusion/models/ideogram/qwen3_text_encoder.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
+import jax
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/qwen3_text_encoder.py:15]: this file and some other files are missing the license header, this would incur errors when doing copy-bara

File: src/maxdiffusion/models/ideogram/qwen3_text_encoder.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/qwen3_text_encoder.py:16]: also this file is named for torchax but there's no torchax here (I see `device=torch.device("cpu")`)

File: src/maxdiffusion/models/ideogram/quantized_loading.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        torch.empty(out_features, in_features, dtype=FP8_WEIGHT_DTYPE),
+    )
+    self.register_buffer("weight_scale", torch.empty(out_features, dtype=torch.float32))
+    if bias:
+      self.register_buffer("bias", torch.empty(out_features, dtype=compute_dtype))
+    else:
+      self.bias = None
+
+  def forward(self, x: torch.Tensor) -> torch.Tensor:
+    w = self.weight.to(x.dtype) * self.weight_scale.to(x.dtype).unsqueeze(1)
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/quantized_loading.py]: this line dequantizes the full weight tensor on every forward. We should dequantize once at load and cache

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+    # Transpose to (B, num_heads, L, head_dim)
+    q = jnp.transpose(q, (0, 2, 1, 3))
+    k = jnp.transpose(k, (0, 2, 1, 3))
+    v = jnp.transpose(v, (0, 2, 1, 3))
+
+    q, k = _apply_rotary_pos_emb(q, k, cos, sin)
+
+    # Block-diagonal mask from segment ids
+    attn_mask = jnp.expand_dims(segment_ids, axis=2) == jnp.expand_dims(segment_ids, axis=1)
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/transformer_ideogram.py:129]: I am not quite sure why we need masks to do this...we can leave this for future optimization

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    attn_mask = jnp.expand_dims(attn_mask, axis=1)  # (B, 1, L, L)
+
+    # JAX scaled dot product attention
+    scale = 1.0 / math.sqrt(self.head_dim)
+    attn_weights = jnp.einsum("bhqd,bhkd->bhqk", q, k) * scale
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/transformer_ideogram.py:134]: in `base_ideogram.yml` we have `attention: "flash" ` but here it uses a naive attention, can be optimized later

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+  cos = jnp.expand_dims(cos, axis=1)
+  sin = jnp.expand_dims(sin, axis=1)
+  q_embed = (q * cos) + (_rotate_half(q) * sin)
+  k_embed = (k * cos) + (_rotate_half(k) * sin)
+  return q_embed, k_embed
+
+
+class Ideogram4Attention(nnx.Module):
+
+  def __init__(self, rngs: nnx.Rngs, hidden_size: int, num_heads: int, eps: float = 1e-5, dtype=jnp.float32):
```
Comment (eltsai) [inline src/maxdiffusion/models/ideogram/transformer_ideogram.py:97]: `dtype=jnp.float32` default is never overridden, so `weights_dtype/activations_dtype: bfloat16` in the config are inert.

File: src/maxdiffusion/pipelines/ideogram/ideogram_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    # Initialize z
+    key = jax.random.PRNGKey(seed)
+    latent_dim = 128  # 32 * patch_size * patch_size
+    z = jax.random.normal(key, (batch_size, num_image_tokens, latent_dim), dtype=jnp.float32)
+
+    # Precompute the scheduler timesteps array using the Ideogram 4 scheduler
+    schedule_fn = get_schedule_for_resolution((height, width), known_mean=0.5)
+    step_intervals = make_step_intervals(num_steps)
+    # Compute schedule evaluated at all intervals. schedule_fn returns a continuous scalar mapping.
+    sigmas = jnp.array(
```
Comment (eltsai) [inline src/maxdiffusion/pipelines/ideogram/ideogram_pipeline.py]: this can be vectorized

File: src/maxdiffusion/generate_ideogram.py (hunk 1 above)
Comment (mbohlool) [reply src/maxdiffusion/generate_ideogram.py]: Addressed: Updated the warmup compilation pass in `generate_ideogram.py` to use `num_inference_steps = original_num_steps` so XLA compiles the full 50-step loop during warmup, removing the compilation overhead from the timed inference pass. But the inference time is is ~50s. We can improve it in a follow up PR.

File: src/maxdiffusion/models/ideogram/qwen3_text_encoder.py (hunk 2 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/qwen3_text_encoder.py:15]: Addressed: Added standard Apache 2.0 license headers across all newly added Ideogram files.

File: src/maxdiffusion/models/ideogram/qwen3_text_encoder.py (hunk 3 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/qwen3_text_encoder.py:16]: Addressed: Renamed the file to `qwen3_text_encoder.py` and the class to `Qwen3VLTextEncoder`. It executes in PyTorch on CPU to handle FP8 quantized weights and multi-layer intermediate hidden state taps.

File: src/maxdiffusion/models/ideogram/quantized_loading.py (hunk 4 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/quantized_loading.py]: Addressed: Added `_cached_weight` on `Fp8Linear` and updated `load_fp8_state_dict` to dequantize all weights once at load time.

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py (hunk 5 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/transformer_ideogram.py:129]: Acknowledged: The segment mask handles variable-length sequence padding vs. active token boundaries. We'll leave specialized splash kernel optimizations for a follow-up PR.

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py (hunk 6 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/transformer_ideogram.py:134]: Acknowledged: We'll follow up in a subsequent PR to add specialized Flash / Splash attention kernel integration for Ideogram.

File: src/maxdiffusion/models/ideogram/transformer_ideogram.py (hunk 7 above)
Comment (mbohlool) [reply src/maxdiffusion/models/ideogram/transformer_ideogram.py:97]: Addressed: Made `dtype` configurable in `Ideogram4Config` and `Ideogram4Transformer`, and ensured `IdeogramPipeline` passes `config.weights_dtype` (`bfloat16`) to the transformer instances.

File: src/maxdiffusion/pipelines/ideogram/ideogram_pipeline.py (hunk 8 above)
Comment (mbohlool) [reply src/maxdiffusion/pipelines/ideogram/ideogram_pipeline.py]: Addressed: Replaced the list comprehension with a vectorized `sigmas = schedule_fn(step_intervals)` call.
